### PR TITLE
803 implement knit print for subclasses of stopping

### PR DESCRIPTION
**Pull request by Puzzled-Face**
`knit_print` for subclasses of `Stopping`.

- Please check my interpretation of the more esoteric stopping rules, particularly `StoppingMTDCV`, `StoppingTDCIRatio` and `StoppingMaxGainCIRatio`.
-  There are no snapshots in the tests for these new methods :-)

### CODE REVIEW

File: tests/testthat/test-helpers_knitr_Stopping.R [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+test_that("knit_print.StoppingMTDCV works correctly", {
+  expect_equal(
+    knit_print(
+      StoppingMTDCV(target = 0.33, thresh = 0.5, report_label = "MTD_CV"),
+      asis = FALSE
+    ),
+    paste0(
+      "MTD_CV: If the posterior estimate of the robust coefficient of variation ",
+      "of the MTD, expressed as a percentage, is both greater than zero and less ",
+      "than or equal to 50%."
```
Comment (danielinteractive) [inline tests/testthat/test-helpers_knitr_Stopping.R:217]:
```suggestion
      "MTD_CV: If the posterior estimate of the robust coefficient of variation ",
      "of the MTD (targeting 33%), is below 50%."
```
(mentioning the `target` of the MTD, and simplifying the condition - the MTD_CV will always be non-negative)

File: tests/testthat/test-helpers_knitr_Stopping.R [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+    )
+  )
+  expect_equal(
+    knit_print(
+      StoppingMTDCV(target = 0.33, thresh = 0.5),
+      asis = FALSE
+    ),
+    paste0(
+      "If the posterior estimate of the robust coefficient of variation ",
```
Comment (danielinteractive) [inline tests/testthat/test-helpers_knitr_Stopping.R:226]: same here

File: tests/testthat/test-helpers_knitr_Stopping.R [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+test_that("knit_print.StoppingTDCIRatio works correctly", {
+  expect_equal(
+    knit_print(
+      StoppingTDCIRatio(target_ratio = 4, prob_target = 0.3, report_label = "TDCI_RATIO"),
+      asis = FALSE
+    ),
+    paste0(
+      "TDCI_RATIO: If, at the next best dose, the ratio of the upper and lower ",
+      "limits of the posterior 95% credible interval for the probability of ",
+      "toxicity is less than or equal to 4."
```
Comment (danielinteractive) [inline tests/testthat/test-helpers_knitr_Stopping.R:140]:
```suggestion
      "TDCI_RATIO: If, at the next best dose, the ratio of the upper to the lower ",
      "limit of the posterior 95% credible interval for the MTD (targeting 30%) ",
      "is less than or equal to 4."
```

File: tests/testthat/test-helpers_knitr_Stopping.R [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+    )
+  )
+  expect_equal(
+    knit_print(
+      StoppingTDCIRatio(target_ratio = 4, prob_target = 0.3),
+      asis = FALSE,
+      dose_label = "the MTD"
+    ),
+    paste0(
+      "If, at the MTD, the ratio of the upper and lower ",
```
Comment (danielinteractive) [inline tests/testthat/test-helpers_knitr_Stopping.R:150]: similar here

File: tests/testthat/test-helpers_knitr_Stopping.R [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+  expect_equal(
+    knit_print(
+      StoppingMaxGainCIRatio(target_ratio = 4, prob_target = 0.3, report_label = "MAXGAINCI_RATIO"),
+      asis = FALSE
+    ),
+    paste0(
+      "MAXGAINCI_RATIO: If the ratio of the upper and lower limits of the ",
+      "posterior 95% credible interval for the probability of toxicity at the ",
+      "smaller of the doses given by the end-of-trial TD estimate and the G* ",
+      "estimate is less than or equal to 4."
```
Comment (danielinteractive) [inline tests/testthat/test-helpers_knitr_Stopping.R:113]:
```suggestion
      "MAXGAINCI_RATIO: If the ratio of the upper to the lower limit of the ",
      "posterior 95% credible interval for the probability of toxicity at the target dose ",
      "(the smaller one of the MTD for 30% target and GStar) ",
      "is less than or equal to 4."
```

File: tests/testthat/test-helpers_knitr_Stopping.R [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+    )
+  )
+  expect_equal(
+    knit_print(
+      StoppingMaxGainCIRatio(target_ratio = 4, prob_target = 0.3),
+      asis = FALSE,
+      dose_label = "the MTD"
+    ),
+    paste0(
+      "If the ratio of the upper and lower limits of the ",
```
Comment (danielinteractive) [inline tests/testthat/test-helpers_knitr_Stopping.R:123]: similar here

**Review by danielinteractive — COMMENTED** [review]
Thanks @Puzzled-Face , please see suggestions below for the esoteric stopping rules

File: tests/testthat/test-helpers_knitr_Stopping.R (hunk 1 above)
Comment (Puzzled-Face) [reply tests/testthat/test-helpers_knitr_Stopping.R:217]: OK.  I thought the CV should always be > 0.  The code does test for CV > 0, though.  
```
do_stop <- (mtd_cv <= stopping@thresh_cv) && (mtd_cv >= 0)
```
Only a small thing, but I can remove if you wish.

Comment (danielinteractive) [reply tests/testthat/test-helpers_knitr_Stopping.R:217]: yeah can remove there

File: tests/testthat/test-helpers_knitr_Stopping.R (hunk 3 above)
Comment (Puzzled-Face) [reply tests/testthat/test-helpers_knitr_Stopping.R:140]: would "... (targeting p(Tox) = 30%) ..." be better?  (And similarly for `StoppingMaxGainCIRatio`...)

Comment (danielinteractive) [reply tests/testthat/test-helpers_knitr_Stopping.R:140]: hm not sure, I would not put too much math notation in there

**Review by danielinteractive — APPROVED** [review]
Thanks @Puzzled-Face !

**Pull request merged**